File: evaluation-observe/deploy-langfuse-on-ecs-with-fargate/langfuse-v3/cdk_stacks/ecr.py
#!/usr/bin/env python3
# -*- encoding: utf-8 -*-
# vim: tabstop=2 shiftwidth=2 softtabstop=2 expandtab

import aws_cdk as cdk

from aws_cdk import (
  Stack,
  aws_ecr,
)
from constructs import Construct
import cdk_ecr_deployment as ecr_deploy


class ECRStack(Stack):

  def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
    super().__init__(scope, construct_id, **kwargs)

    ecr_repositories = self.node.try_get_context('ecr')

    self.repositories = {}

    for elem in ecr_repositories:
      repository_name = elem['repository_name']

      repository = aws_ecr.Repository(self, f"{repository_name}ECRRepository",
        empty_on_delete=True,
        encryption=aws_ecr.RepositoryEncryption.AES_256,
        removal_policy=cdk.RemovalPolicy.DESTROY,
        repository_name=repository_name
      )
      self.repositories[repository_name] = repository

      # delete images older than 7 days
      repository.add_lifecycle_rule(
        max_image_age=cdk.Duration.days(7),
        rule_priority=1,
        tag_status=aws_ecr.TagStatus.UNTAGGED
      )

      # keep last 3 images
      repository.add_lifecycle_rule(
        max_image_count=3,
        rule_priority=2,
        tag_status=aws_ecr.TagStatus.ANY
      )

      src_docker_image_version = elem['docker_image_name']
      image_version = elem.get('tag', 'latest')
      deploy_image_versions = image_version if image_version == "latest" else [image_version, "latest"]

      for idx, deploy_image_version in enumerate(deploy_image_versions):
        ecr_deploy.ECRDeployment(self, f"{repository_name}ECRDeployment-{idx:0>3}",
          src=ecr_deploy.DockerImageName(f'{src_docker_image_version}:{image_version}'),
          dest=ecr_deploy.DockerImageName(repository.repository_uri_for_tag_or_digest(deploy_image_version))
        )

File: evaluation-observe/deploy-langfuse-on-ecs-fargate-with-typescript-cdk/lib/langfuse/ecr.ts
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0
// External Dependencies:
import * as cdk from "aws-cdk-lib";
import * as ecr from "aws-cdk-lib/aws-ecr";
import * as ecrDeploy from "cdk-ecr-deployment";
import { NagSuppressions } from "cdk-nag";
import { Construct } from "constructs";

export interface IECRRepoAndDockerImageProps {
  /**
   * Name of the pre-existing Docker image to stage, including tag
   *
   * e.g. 'clickhouse:25.1' or 'langfuse/langfuse:latest'
   */
  dockerImageName: string;
  /**
   * Tag to give the *staged* image in ECR, e.g. 'latest' or '1.0.0'
   *
   * @default 'latest'
   */
  ecrImageTag?: string;
  /**
   * Name for the ECR image repository.
   *
   * The repository name must start with a letter and can only contain lowercase letters, numbers,
   * hyphens, underscores, and forward slashes. If you specify a name, you cannot perform updates
   * that require replacement of this resource. You can perform updates that require no or some
   * interruption. If you must replace the resource, specify a new name.
   *
   * @default undefined Automatically generated name.
   */
  repositoryName?: string;
  /**
   * AWS Tags to apply to created image registry
   */
  tags?: cdk.Tag[];
}

/**
 * Construct to create an ECR repository with sensible settings and stage a Docker image to it
 */
export class ECRRepoAndDockerImage extends Construct {
  public readonly repository: ecr.Repository;
  public readonly deployment: ecrDeploy.ECRDeployment;
  public readonly imageTag: string;

  constructor(
    scope: Construct,
    id: string,
    props: IECRRepoAndDockerImageProps,
  ) {
    super(scope, id);

    // Create the ECR repository:
    this.repository = new ecr.Repository(this, "Repo", {
      emptyOnDelete: true,
      encryption: ecr.RepositoryEncryption.AES_256,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      repositoryName: props.repositoryName,
    });
    if (props.tags) {
      props.tags.forEach((tag) =>
        cdk.Tags.of(this.repository).add(tag.key, tag.value),
      );
    }

    // Automatically clean up old or un-tagged images:
    this.repository.addLifecycleRule({
      maxImageAge: cdk.Duration.days(7),
      rulePriority: 1,
      tagStatus: ecr.TagStatus.UNTAGGED,
    });
    this.repository.addLifecycleRule({
      maxImageCount: 3,
      rulePriority: 2,
      tagStatus: ecr.TagStatus.ANY,
    });

    // Add the target image version(s):
    this.imageTag = props.ecrImageTag || "latest";
    this.deployment = new ecrDeploy.ECRDeployment(this, "Deployment", {
      src: new ecrDeploy.DockerImageName(props.dockerImageName),
      dest: new ecrDeploy.DockerImageName(
        this.repository.repositoryUriForTagOrDigest(this.imageTag),
      ),
    });

    const stack = cdk.Stack.of(this);
    NagSuppressions.addResourceSuppressionsByPath(
      stack,
      `${stack.node.path}/Custom::CDKECRDeploymentbd07c930edb94112a20f03f096f53666512MiB/ServiceRole/Resource`,
      [
        {
          id: "AwsSolutions-IAM4",
          reason:
            "Can't control managed policies within ECRDeployment resource",
        },
      ],
    );
    NagSuppressions.addResourceSuppressionsByPath(
      stack,
      `${stack.node.path}/Custom::CDKECRDeploymentbd07c930edb94112a20f03f096f53666512MiB/ServiceRole/DefaultPolicy/Resource`,
      [
        {
          id: "AwsSolutions-IAM5",
          reason:
            "Can't control wildcard policies within ECRDeployment resource",
        },
      ],
    );
  }
}
